cibuild.sh

```diff
@@ -0,0 +1,171 @@
+path+=("$prebuilt/gcc-arm-none-eabi/bin")
+
+# mips gcc toolchain
+install+=("if [ ! -d \"$prebuilt/pinguino-compilers/linux64/p32/bin\" ]; then \
+             cd $prebuilt; \
+             git clone https://github.com/PinguinoIDE/pinguino-compilers; \
+           fi; \
+           p32-gcc --version; \
+           ")
+path+=("$prebuilt/pinguino-compilers/linux64/p32/bin")
+
+# risc-v gcc toolchain
+install+=("if [ ! -d \"$prebuilt/riscv64-unknown-elf-gcc/bin\" ]; then \
+             cd $prebuilt; \
+             wget https://static.dev.sifive.com/dev-tools/riscv64-unknown-elf-gcc-8.3.0-2019.08.0-x86_64-linux-ubuntu14.tar.gz; \
+             tar zxf riscv64-unknown-elf-gcc-8.3.0-2019.08.0-x86_64-linux-ubuntu14.tar.gz; \
+             mv riscv64-unknown-elf-gcc-8.3.0-2019.08.0-x86_64-linux-ubuntu14 riscv64-unknown-elf-gcc; \
+             rm riscv64-unknown-elf-gcc-8.3.0-2019.08.0-x86_64-linux-ubuntu14.tar.gz; \
+           fi; \
+           riscv64-unknown-elf-gcc --version; \
+           ")
+path+=("$prebuilt/riscv64-unknown-elf-gcc/bin")
+
+# ccache
+install+=("if [ ! -d \"$prebuilt/ccache/bin\" ]; then \
+             cd $prebuilt; \
+             wget https://github.com/ccache/ccache/releases/download/v3.7.7/ccache-3.7.7.tar.gz; \
+             tar zxf ccache-3.7.7.tar.gz; \
+             cd ccache-3.7.7; ./configure --prefix=$prebuilt/ccache; make; make install; \
+             cd $prebuilt; rm -rf ccache-3.7.7; rm ccache-3.7.7.tar.gz; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/gcc; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/g++; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/arm-none-eabi-gcc; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/arm-none-eabi-g++; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/p32-gcc; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/riscv64-unknown-elf-gcc; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/riscv64-unknown-elf-g++; \
+           fi; \
+           ccache --version; \
+           export USE_CCACHE=1; \
+           export CCACHE_DIR=$prebuilt/ccache/.ccache; \
+           ccache -M 5G; \
+           ")
+path+=("$prebuilt/ccache/bin")
+
+function usage {
+  echo ""
+  echo "USAGE: $0 check | full"
+  echo "       $0 -h"
+  echo ""
+  echo "Where:"
+  echo "  check:   do check build CI"
+  echo "  full:    do full build CI"
+  echo "  -h will show this help test and terminate"
+  echo ""
+  exit 1
+}
+
+function setup_repos {
+  if [ -d "$nuttx" ]; then
+    cd $nuttx; git pull
+  else
+    git clone https://github.com/apache/incubator-nuttx.git $nuttx
+  fi
+
+  if [ -d "$apps" ]; then
+    cd $apps; git pull
+  else
+    git clone https://github.com/apache/incubator-nuttx-apps.git $apps
+  fi
+}
+
+function install_tools {
+  for ((i=0; i<${#install[@]}; ++i)); do
+    export PATH=${path[$i]}:$PATH
+    eval ${install[$i]}
```

Comment:
Using `eval` on a long strings of code makes difficult to edit (editors will not highlight the code inside the literal string). To me it looks like extra complexity for no apparent benefit. Why not putting directly the code like in `install_tools()` or at least define function for each one, and then call them from `install_tools()`?.

Reply:
In fact, with set -e and set -x in bash, it's not much difficult to debug.  IMO, add tools into an array may be clear than more functions. Once tools need update, just update the array.

Reply:
The fact that `-x` helps with that problem should not be a reason if it can be avoided. And if you prefer to control the tools installed using an array it can be done also using functions. Also since the functions have names, the execution with `-x` gives extra info. For example: 
```
#!/usr/bin/env bash
set -x
set -e

PATH="initial_path"

function add_path {
    PATH=$1:$PATH
}

function tool1 {
    echo installing tool 1
    add_path "/path/to/tool1"
}

function tool2 { 
    echo installing tool 2
    add_path "/path/to/tool2"
}

install="tool1 tool2"

# ... and a few lines later ...

for f in $install; do
    $f
done
```
will output
```
+ set -e
+ PATH=initial_path
+ install='tool1 tool2'
+ for f in $install
+ tool1
+ echo installing tool 1
installing tool 1
+ add_path /path/to/tool1
+ PATH=/path/to/tool1:initial_path
+ for f in $install
+ tool2
+ echo installing tool 2
installing tool 2
+ add_path /path/to/tool2
+ PATH=/path/to/tool2:/path/to/tool1:initial_path
```

Reply:
Thanks, it makes sense. update on going.

Reply:
done, maht help to review, thanks.